### PR TITLE
[bugfix] fix transport race conditions in yurthub

### LINKED ISSUE #674: [BUG] Race conditions in yurthub
**What happened**:
The `rp.reverseProxy` in the below codes will be modified by multiple goroutines without any synchronization mechanism, which leads to race conditions.
https://github.com/openyurtio/openyurt/blob/5063752a9f6645270e3177e39a46df8c23145af2/pkg/yurthub/proxy/remote/remote.go#L91-L105

I built an image with a race detector enabled binary (go build -race )  of yurthub, and it verified my thought:
```
I1008 05:40:00.037350       1 reflector.go:175] Starting reflector *v1alpha1.NodePool (24h0m0s) from pkg/mod/k8s.io/client-go@v0.18.8/tools/cache/reflector.go:125
==================
WARNING: DATA RACE
Write at 0x00c000878f08 by goroutine 126:
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/remote.(*RemoteProxy).ServeHTTP()
      /root/openyurt/pkg/yurthub/proxy/remote/remote.go:92 +0xb0
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/remote.(*loadBalancer).ServeHTTP()
      /root/openyurt/pkg/yurthub/proxy/remote/loadbalancer.go:189 +0x325
  github.com/openyurtio/openyurt/pkg/yurthub/proxy.(*yurtReverseProxy).ServeHTTP()
      /root/openyurt/pkg/yurthub/proxy/proxy.go:115 +0x18c
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithRequestTrace.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:213 +0x2c5
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithRequestContentType.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:71 +0x14a
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithCacheHeaderCheck.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:92 +0x14a
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithRequestTimeout.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:288 +0x14a
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithListRequestSelector.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:139 +0x152
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithMaxInFlightLimit.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:233 +0x386
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithRequestClientComponent.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:165 +0x14a
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1()
      /root/go/pkg/mod/k8s.io/apiserver@v0.18.8/pkg/endpoints/filters/requestinfo.go:39 +0x3fc
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  net/http.serverHandler.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2843 +0xca
  net/http.(*conn).serve()
      /usr/lib/golang/src/net/http/server.go:1925 +0x84c

Previous write at 0x00c000878f08 by goroutine 127:
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/remote.(*RemoteProxy).ServeHTTP()
      /root/openyurt/pkg/yurthub/proxy/remote/remote.go:92 +0xb0
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/remote.(*loadBalancer).ServeHTTP()
      /root/openyurt/pkg/yurthub/proxy/remote/loadbalancer.go:189 +0x325
  github.com/openyurtio/openyurt/pkg/yurthub/proxy.(*yurtReverseProxy).ServeHTTP()
      /root/openyurt/pkg/yurthub/proxy/proxy.go:115 +0x18c
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithRequestTrace.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:213 +0x2c5
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithRequestContentType.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:71 +0x14a
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithCacheHeaderCheck.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:92 +0x14a
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithRequestTimeout.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:288 +0x14a
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithListRequestSelector.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:139 +0x152
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithMaxInFlightLimit.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:233 +0x386
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  github.com/openyurtio/openyurt/pkg/yurthub/proxy/util.WithRequestClientComponent.func1()
      /root/openyurt/pkg/yurthub/proxy/util/util.go:165 +0x14a
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  k8s.io/apiserver/pkg/endpoints/filters.WithRequestInfo.func1()
      /root/go/pkg/mod/k8s.io/apiserver@v0.18.8/pkg/endpoints/filters/requestinfo.go:39 +0x3fc
  net/http.HandlerFunc.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2042 +0x51
  net/http.serverHandler.ServeHTTP()
      /usr/lib/golang/src/net/http/server.go:2843 +0xca
  net/http.(*conn).serve()
      /usr/lib/golang/src/net/http/server.go:1925 +0x84c

Goroutine 126 (running) created at:
  net/http.(*Server).Serve()
      /usr/lib/golang/src/net/http/server.go:2969 +0x5d3
  net/http.(*Server).ListenAndServe()
      /usr/lib/golang/src/net/http/server.go:2866 +0x105
  github.com/openyurtio/openyurt/pkg/yurthub/server.(*yurtHubServer).Run()
      /root/openyurt/pkg/yurthub/server/server.go:143 +0xb6
  github.com/openyurtio/openyurt/cmd/yurthub/app.Run()
      /root/openyurt/cmd/yurthub/app/start.go:184 +0x2195
  github.com/openyurtio/openyurt/cmd/yurthub/app.NewCmdStartYurtHub.func1()
      /root/openyurt/cmd/yurthub/app/start.go:71 +0x584
  github.com/spf13/cobra.(*Command).execute()
      /root/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846 +0x90c
  github.com/spf13/cobra.(*Command).ExecuteC()
      /root/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x4b2
  github.com/spf13/cobra.(*Command).Execute()
      /root/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887 +0x146
  main.main()
      /root/openyurt/cmd/yurthub/yurthub.go:33 +0x151

Goroutine 127 (running) created at:
  net/http.(*Server).Serve()
      /usr/lib/golang/src/net/http/server.go:2969 +0x5d3
  net/http.(*Server).ListenAndServe()
      /usr/lib/golang/src/net/http/server.go:2866 +0x105
  github.com/openyurtio/openyurt/pkg/yurthub/server.(*yurtHubServer).Run()
      /root/openyurt/pkg/yurthub/server/server.go:143 +0xb6
  github.com/openyurtio/openyurt/cmd/yurthub/app.Run()
      /root/openyurt/cmd/yurthub/app/start.go:184 +0x2195
  github.com/openyurtio/openyurt/cmd/yurthub/app.NewCmdStartYurtHub.func1()
      /root/openyurt/cmd/yurthub/app/start.go:71 +0x584
  github.com/spf13/cobra.(*Command).execute()
      /root/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:846 +0x90c
  github.com/spf13/cobra.(*Command).ExecuteC()
      /root/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x4b2
  github.com/spf13/cobra.(*Command).Execute()
      /root/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887 +0x146
  main.main()
      /root/openyurt/cmd/yurthub/yurthub.go:33 +0x151
==================
```

To fix that, I think we should group all `Transport` into one and run different logic for corresponding requests.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- OpenYurt version:
- Kubernetes version (use `kubectl version`):
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

**others**

/kind bug

### PR DESCRIPTION
**Pull request by rambohe-ch**
/kind bug

#### What this PR does / why we need it:
reconstruct transport setting in `RemoteProxy` of yurthub component for preventing race conditions.
1.  add `bearerUpgradeHandler` and existing `upgradeHandler` for `Connection: Upgrade` request.
2. add `RoundTrip()` func for `RemoteProxy` to group transport handler.

#### Which issue(s) this PR fixes:

Fixes #674 

#### Special notes for your reviewer:

/assign @DrmagicE 

#### Does this PR introduce a user-facing change?

```release-note

```

### CODE REVIEW

File: pkg/yurthub/proxy/remote/remote.go [hunk 1]
```diff
@@ -74,22 +75,25 @@ func NewRemoteProxy(remoteServer *url.URL,
 		return nil, fmt.Errorf("could not get bearer transport when init proxy backend(%s)", remoteServer.String())
 	}
 
-	upgradeAwareHandler := proxy.NewUpgradeAwareHandler(remoteServer, nil, false, true, &responder{})
+	upgradeAwareHandler := proxy.NewUpgradeAwareHandler(remoteServer, currentTransport, false, true, &responder{})
 	upgradeAwareHandler.UseRequestLocation = true
+	bearerUpgradeAwareHandler := proxy.NewUpgradeAwareHandler(remoteServer, bearerTransport, false, true, &responder{})
```
[the remark this answers is not in the document]
Comment (rambohe-ch) [reply pkg/yurthub/proxy/remote/remote.go:80]: @Congrool From the code: https://github.com/kubernetes/apimachinery/blob/v0.18.8/pkg/util/proxy/upgradeaware.go#L398-L400

when UpgradeTransport is nil, UpgradeAwareHandler will use Transport to handle upgrade request.

File: pkg/yurthub/proxy/remote/remote.go [hunk 2]
```diff
@@ -106,21 +110,13 @@ func (rp *RemoteProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	if httpstream.IsUpgradeRequest(req) {
 		klog.V(5).Infof("get upgrade request %s", req.URL)
 		if isBearerRequest(req) {
-			rp.upgradeHandler.UpgradeTransport = proxy.NewUpgradeRequestRoundTripper(rp.bearerTransport, proxy.MirrorRequest)
+			rp.bearerUpgradeHandler.ServeHTTP(rw, req)
 		} else {
-			rp.upgradeHandler.UpgradeTransport = proxy.NewUpgradeRequestRoundTripper(rp.currentTransport, proxy.MirrorRequest)
+			rp.upgradeHandler.ServeHTTP(rw, req)
```
[the remark this answers is not in the document]
Comment (rambohe-ch) [reply pkg/yurthub/proxy/remote/remote.go:115]: the same as above

File: pkg/yurthub/proxy/remote/remote.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (DrmagicE) [reply pkg/yurthub/proxy/remote/remote.go:80]: I think we can group `upgradeAwareHandler` and `bearerUpgradeAwareHandler` into one. 

`upgradeAwareHandler` and `bearerUpgradeAwareHandler` here are used to distinguish bearer and non-bearer requests and use different transport for them.  The `RemoteProxy` is exactly what we need:
```
func (rp *RemoteProxy) RoundTrip(req *http.Request) (*http.Response, error) {
	// when edge client(like kube-proxy, flannel, etc) use service account(default InClusterConfig) to access yurthub,
	// Authorization header will be set in request. and when edge client(like kubelet) use x509 certificate to access
	// yurthub, Authorization header in request will be empty.
	if isBearerRequest(req) {
		return rp.bearerTransport.RoundTrip(req)
	}

	return rp.currentTransport.RoundTrip(req)
}
```

I think we can remove `bearerUpgradeAwareHandler` and use `proxyBackend` as the transport param in `NewUpgradeAwareHandler`.  

The codes should look like:
```
	proxyBackend := &RemoteProxy{
		checker:          healthChecker,
		reverseProxy:     httputil.NewSingleHostReverseProxy(remoteServer),
		cacheMgr:         cacheMgr,
		remoteServer:     remoteServer,
		filterChain:      filterChain,
		currentTransport: currentTransport,
		bearerTransport:  bearerTransport,
		stopCh:           stopCh,
	}

	upgradeAwareHandler := proxy.NewUpgradeAwareHandler(remoteServer, proxyBackend, false, true, &responder{})
	upgradeAwareHandler.UseRequestLocation = true

	proxyBackend.reverseProxy.Transport = proxyBackend
	proxyBackend.reverseProxy.ModifyResponse = proxyBackend.modifyResponse
	proxyBackend.reverseProxy.FlushInterval = -1
	proxyBackend.reverseProxy.ErrorHandler = proxyBackend.errorHandler
	proxyBackend.upgradeHandler = upgradeAwareHandler
```

File: pkg/yurthub/proxy/remote/remote.go [hunk 3]
```diff
@@ -106,21 +110,13 @@ func (rp *RemoteProxy) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	if httpstream.IsUpgradeRequest(req) {
 		klog.V(5).Infof("get upgrade request %s", req.URL)
 		if isBearerRequest(req) {
```
Comment (DrmagicE) [inline pkg/yurthub/proxy/remote/remote.go:112]: As mentioned in https://github.com/openyurtio/openyurt/pull/683/files#r769771572,  if we group `upgradeAwareHandler` and `bearerUpgradeAwareHandler` into one, the codes here can be simplified as follow:
```
	if httpstream.IsUpgradeRequest(req) {
		klog.V(5).Infof("get upgrade request %s", req.URL)
		rp.upgradeHandler.ServeHTTP(rw, req)
		return
	}

	rp.reverseProxy.ServeHTTP(rw, req)
```

File: pkg/yurthub/proxy/remote/remote.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rambohe-ch) [reply pkg/yurthub/proxy/remote/remote.go:80]: @DrmagicE A good idea, i will modify it.

[the remark this answers is not in the document]
Comment (rambohe-ch) [reply pkg/yurthub/proxy/remote/remote.go:80]: @DrmagicE I have checked the codes of `NewUpgradeAwareHandler`, and only `*http.Transport` as `http. RoundTripper` is supported, so use proxyBackend as `http. RoundTripper` can not work. 
maybe we need to leave the current changed unchanged.

File: pkg/yurthub/proxy/remote/remote.go (hunk 3 above)
Comment (rambohe-ch) [reply pkg/yurthub/proxy/remote/remote.go:112]: As replied above

File: pkg/yurthub/proxy/remote/remote.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (DrmagicE) [reply pkg/yurthub/proxy/remote/remote.go:80]: @rambohe-ch Oh, what a pity. I got it.

**Pull request merged**